File: repofactory/src/main/java/com/guidovezzoni/architecture/datasource/RetrofitFunctionDataSource.java
package com.guidovezzoni.architecture.datasource;

import com.fernandocejas.arrow.checks.Preconditions;
import com.guidovezzoni.architecture.cache.TimeStampHelper;
import io.reactivex.Single;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import retrofit2.Response;

public class RetrofitFunctionDataSource<M, P> extends BaseRetrofitDataSource<M, P> {
    @NotNull
    private final Function1<P, Single<Response<M>>> endPointGet;

    public RetrofitFunctionDataSource(@NotNull TimeStampHelper timeStampHelper, @NotNull Function1<P, Single<Response<M>>> endPointGet) {
        super(timeStampHelper);
        Preconditions.checkNotNull(endPointGet);
        this.endPointGet = endPointGet;
    }

    @Override
    protected Single<Response<M>> getFromEndPoint(P params) {
        return endPointGet.invoke(params);
    }
}

File: repofactory/src/main/java/com/guidovezzoni/architecture/datasource/RetrofitFunctionDataSource.kt
package com.guidovezzoni.architecture.datasource

import com.guidovezzoni.architecture.cache.TimeStampHelper
import com.guidovezzoni.utils.RetrofitUtils
import io.reactivex.Single
import io.reactivex.SingleTransformer
import retrofit2.Response

/**
 * implementation of [BaseRetrofitDataSource] that uses a [Function1] to define [getFromEndPoint]
 */
open class RetrofitFunctionDataSource<M, P> @JvmOverloads constructor(
    timeStampHelper: TimeStampHelper,
    retrofitUnwrapper: SingleTransformer<Response<M>, M> = RetrofitUtils.unWrapResponse(),
    private val endPointGet: Function1<P?, Single<Response<M>>>
) : BaseRetrofitDataSource<M, P>(timeStampHelper, retrofitUnwrapper) {

    public override fun getFromEndPoint(params: P?): Single<Response<M>> {
        return endPointGet.invoke(params)
    }
}
